### PR TITLE
Ticket 284 fix paractrystal

### LINKED ISSUE #284: dnn should be lattice spacing for paracrystalline models (Trac #1156)
[Copied from ticket comment:4:ticket:805]

The parameter to the model is labelled as nearest neighbour distance, but I believe it is lattice spacing.

In units of lattice spacing, a, the q positions of the peaks are well determined by:

   q = 2*pi/a * sqrt(h^2 + k^2 + l^2)

where SC allows all integer hkl, BCC requires h+k+l even and FCC require hkl all even or hkl all odd. Also need peak multiplicity to predict the peak height, since ```[100] = [-100] = [010] = [0-10] = [001] = [00-1]```, etc.

I checked peak position and height for dnn=20 and d_factor=0.06 for sc, bcc and fcc to verify the calculation.

We should rename dnn to length_a as in the parallelepiped model, and rename d_factor to lattice_distortion. Lattice distortion is actually proportional, Delta a / a, but distortion_proportion is too much of a mouthful.

Migrated from http://trac.sasview.org/ticket/1156
```json
{
    "status": "assigned",
    "changetime": "2018-11-28T01:44:26",
    "_ts": "2018-11-28 01:44:26.486641+00:00",
    "description": "[Copied from ticket comment:4:ticket:805]\n\nThe parameter to the model is labelled as nearest neighbour distance, but I believe it is lattice spacing.\n\nIn units of lattice spacing, a, the q positions of the peaks are well determined by:\n\n   q = 2*pi/a * sqrt(h^2 + k^2 + l^2)\n\nwhere SC allows all integer hkl, BCC requires h+k+l even and FCC require hkl all even or hkl all odd. Also need peak multiplicity to predict the peak height, since {{{[100] = [-100] = [010] = [0-10] = [001] = [00-1]}}}, etc.\n\nI checked peak position and height for dnn=20 and d_factor=0.06 for sc, bcc and fcc to verify the calculation.\n\nWe should rename dnn to length_a as in the parallelepiped model, and rename d_factor to lattice_distortion. Lattice distortion is actually proportional, Delta a / a, but distortion_proportion is too much of a mouthful.",
    "reporter": "pkienzle",
    "cc": "",
    "resolution": "",
    "workpackage": "SasView Bug Fixing",
    "time": "2018-08-14T15:05:56",
    "component": "SasView",
    "summary": "dnn should be lattice spacing for paracrystalline models",
    "priority": "critical",
    "keywords": "",
    "milestone": "SasView 4.3.0",
    "owner": "butler",
    "type": "defect"
}
```

### LINKED ISSUE #384: paracrystal models return NaN when lattice distortion d_factor is zero
Paracrystalline models fail when the lattice distortion factor is too small, returning NaN values for some or all of the input Q values.

This should be noted in the documentation.

Maybe switch to a different algorithm for small distortions, or set the minimum value of the parameter so that it does not return bad values for reasonable Q.

Check that the Qt GUI handles NaN values in the plots, preferably by warning the user that the theory is bad rather than silently ignoring them.

### PR DESCRIPTION
**Pull request by butlerpd**
This should fix the paracrystal model. I would still like to add the write up by @JoGaudet to the documentation before merging but the maths should be correct now.

Closes #284
Closes #384

### CODE REVIEW

**Review by JoGaudet — APPROVED** [review]

**Review by JoGaudet — APPROVED** [review]

File: sasmodels/models/bcc_paracrystal.py [hunk 1]
```diff
@@ -18,47 +13,105 @@
...
+In this case, the volume fraction ($V_{latt}$) of spherical particles with
+radius $R$ sitting on the bcc lattice is given by:
 
 .. math::
+    V_{latt} &= \frac{4/3 \pi R^3}{a^3/2}\\
+    &= \frac{8\pi R^3}{3a^3}\\
+    &= \frac{\sqrt{3} \pi R^3}{D^3}
 
-    \Delta a = g D
+Now, continuing to follow [#Matsuoka1987]_, we define the structure(lattice)
```
Comment (smk78) [inline sasmodels/models/bcc_paracrystal.py:64]: Insert space between 'structure' and '(lattice)'.

File: sasmodels/models/bcc_paracrystal.py [hunk 2]
```diff
@@ -18,47 +13,105 @@
...
+Here, $\vec{a_k}$ are the primitive unit cell vectors $\vec{a_1}$, $\vec{a_2}$,
+and $\vec{a_3}$. Furthermore, $\Delta a_k$ is the isotropic distortion of the
+lattice point from its ideal position and can be defined by a constant factor
+$g=\Delta a / |\vec{a_1}| = \Delta a / |\vec{a_2}| = \Delta a / |\vec{a_3}|=\Delta a/D$.
+
+Finally, assuming the definitions presented in this document, the authors of
+reference [#Matsuoka1987]_ have derived the lattice factors which are given by:
 
 .. math::
+    Z_1(q,\theta,\phi)&=[1-e^{-q^2\Delta a^2}]/\{1-2e^{-\frac{1}{2}q^2\Delta a^2}cos[\frac{qa}{2}(sin\theta cos\phi + sin\theta sin\phi + cos\theta)] + e^{-q^2\Delta a^2}\}\\
```
Comment (smk78) [inline sasmodels/models/bcc_paracrystal.py:92]: The 'sincos' and 'sinsin' in these definitions of Z1, Z2 & Z3 look ugly. Maybe separate with spaces?

**Review by smk78 — CHANGES_REQUESTED** [review]
See comments

**Comment by smk78** [comment]
Also:
 - in the Note at the end, is it worth suggesting that a GPU be used if available?
 - and I notice in the code that dnn seems to have become d_a; does the parameter dnn therefore need to be renamed?

**Comment by smk78** [comment]
Oh, and there is a typo in the name of this issue! ;-)

File: sasmodels/models/bcc_paracrystal.py (hunk 2 above)
Comment (butlerpd) [reply sasmodels/models/bcc_paracrystal.py:96]: That is standard latex output which I'm not sure how to override.  Figure we can let the latex snobs decide what spacing looks "proper"  :-)

**Comment by butlerpd** [comment]
This should now be ready for final review - docs have been updated for FCC and SC and the reviewer comments addressed.

File: sasmodels/models/bcc_paracrystal.py (hunk 2 above)
Comment (llimeht) [reply sasmodels/models/bcc_paracrystal.py:97]: Would expect the trig functions to be `\sin` and `\cos` in there. Without that, they get rendered as _s_ multiplied by _i_ multiplied by _n_, which certainly is ugly.

Comment (butlerpd) [reply sasmodels/models/bcc_paracrystal.py:97]: oops!  yeah-- thanks @llimeht

**Comment by butlerpd** [comment]
regarding suggesting use of GPUs, the flag `single` is set to `False` which I think means it will not run on the GPUs? In which case suggesting it would just be mean :-)

**Review by smk78 — APPROVED** [review]
These three paracrystal docs now look good to me. Disclaimer: I've scanned the math, but not checked it! :-)

**Comment by smk78** [comment]
> regarding suggesting use of GPUs, the flag `single` is set to `False` which I think means it will not run on the GPUs? In which case suggesting it would just be mean :-)

_single = False_ just means the model requires double-precision. Thus if a dp-GPU, with suitable OpenCL support, were available these models should run on a GPU. The Note in these docs was therefore updated accordingly. If the flag _opencl = False_ that would be more of a problem...

**Comment by butlerpd** [comment]
According to @JoGaudet the SC equations are incorrect and need fixing before this can be merged

**Comment by butlerpd** [comment]
> _single = False_ just means the model requires double-precision. Thus if a dp-GPU, with suitable OpenCL support, were available these models should run on a GPU. The Note in these docs was therefore updated accordingly. If the flag _opencl = False_ that would be more of a problem...

Ok ... according to the docs (as @pkienzle pointed out to me)  _opencl = False_ is the flag for using the GPU.  So yes, the code should run faster with GPU on if it is a "high end" GPU that supports double precision.  Good catch @smk78

**Pull request merged**